### PR TITLE
HF: Several changes to policy assets and getrawtransaction rpc

### LINKED ISSUE #117: RPC: getrawtransaction should show more info
For each transaction output the additional fields will need to be added :

- [x] If asset is policy asset, add an additional field with the asset label

- [x] If the scriptPubkey is of type `lockedmultisig`, try parsing a Request/Bid and add Request/Bid information (similar to getrequests/getreqeustbids rpc)

- [x] extend qa tests where policy assets are used to check that getrawtransaction works as expected

- [x] issuance asset changes

For whitelisting the type is already set to `registeraddress` and data is encrypted. Would we want to show anything else @lawlawlaw ?

@tomt1664 do we want to show anything else for freeze/burn either than the asset label?

### PR DESCRIPTION
**Pull request by AstryX**
I have made significant changes to the getrawtransaction rpc. A new assetlabel field has been added for all policy assets (WHITELIST, BURNLIST, etc). Additionally, an extra field ("request" or "bid") has been added to outputs of lockedmultisig type. These fields parse and store data for transactions that carry information relating to the auction. I have implemented issuanceAsset which has replaced policyAsset in some locations. Changes have been made to QA tests to accommodate these features.

Fixes #117

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -203,25 +219,65 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
...
+        if (asset.IsExplicit()) {
+            if (pindex != NULL) {
+                if (chainActive.Contains(pindex)) {
+                    vector<vector<unsigned char>> vSolutions;
+                    txnouttype whichType;
+                    bool solverRes = Solver(txout.scriptPubKey, whichType, vSolutions);
+                    bool isPolicy = IsPolicy(asset.GetAsset());
+                    if (solverRes && whichType == TX_LOCKED_MULTISIG && isPolicy == true) {
+                        CRequest request = CRequest::FromSolutions(vSolutions, pindex->nHeight);
+                        UniValue item(UniValue::VOBJ);
```
Comment (nkostoulas) [inline src/rpc/rawtransaction.cpp]: You can use the requestToJson method for this

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -203,25 +219,65 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
...
+                        item.push_back(Pair("numTickets", (int32_t)request.nNumTickets));
+                        item.push_back(Pair("decayConst", (int32_t)request.nDecayConst));
+                        item.push_back(Pair("feePercentage", (int32_t)request.nFeePercentage));
+                        item.push_back(Pair("endBlockHeight", (int32_t)request.nEndBlockHeight));
+                        item.push_back(Pair("startPrice", ValueFromAmount(request.nStartPrice)));
+                        item.push_back(Pair("auctionPrice", ValueFromAmount(request.GetAuctionPrice(chainActive.Height()))));
+                        out.push_back(Pair("request", item));
+                    }
+                    else if (solverRes && whichType == TX_LOCKED_MULTISIG && isPolicy == false) {
+                        CBid bid = CBid::FromSolutions(vSolutions, txout.nValue.GetAmount(), pindex->nHeight);
```
Comment (nkostoulas) [inline src/rpc/rawtransaction.cpp]: And the bidToJson method for this

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -203,25 +219,65 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
...
         ScriptPubKeyToJSON(txout.scriptPubKey, o, true);
         out.push_back(Pair("scriptPubKey", o));
+
+        if (asset.IsExplicit()) {
+            if (pindex != NULL) {
+                if (chainActive.Contains(pindex)) {
+                    vector<vector<unsigned char>> vSolutions;
+                    txnouttype whichType;
+                    bool solverRes = Solver(txout.scriptPubKey, whichType, vSolutions);
+                    bool isPolicy = IsPolicy(asset.GetAsset());
```
Comment (nkostoulas) [inline src/rpc/rawtransaction.cpp]: We only care about permissionAsset not all policy asserts for Request transactions.

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -167,6 +178,11 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
             } else if (issuance.nInflationKeys.IsCommitment()) {
                 issue.push_back(Pair("tokenamountcommitment", HexStr(issuance.nInflationKeys.vchCommitment)));
             }
+            const std::string policyLabel = gAssetsDir.GetLabel(asset);
+            if (policyLabel != "") 
+            {
```
Comment (nkostoulas) [inline src/rpc/rawtransaction.cpp]: bracket lint

File: src/rpc/rawtransaction.cpp [hunk 5]
```diff
@@ -203,25 +219,65 @@ void TxToJSON(const CTransaction& tx, const uint256 hashBlock, UniValue& entry)
         } else if (asset.IsCommitment()) {
             out.push_back(Pair("assetcommitment", HexStr(asset.vchCommitment)));
         }
+        if (asset.IsExplicit()) {
+            std::string policyLabel = gAssetsDir.GetLabel(asset.GetAsset());
+            if (policyLabel != "")
```
Comment (nkostoulas) [inline src/rpc/rawtransaction.cpp]: bracket lint

File: src/init.cpp [hunk 6]
```diff
@@ -527,12 +527,14 @@ std::string HelpMessage(HelpMessageMode mode)
             " " + _(" This creates a new chain with a different genesis block."));
         strUsage += HelpMessageOpt("-peginconfirmationdepth", strprintf(_("Pegin claims must be this deep to be considered valid. (default: %d)"), DEFAULT_PEGIN_CONFIRMATION_DEPTH));
         strUsage += HelpMessageOpt("-initialfreecoins", strprintf(_("The amount of OP_TRUE coins created in the genesis block. Primarily for testing. (default: %d)"), 0));
+        strUsage += HelpMessageOpt("-policycoins", strprintf(_("The amount of policy coins created in the genesis block. Primarily for testing. (default: %d)"), 0));
```
Comment (nkostoulas) [inline src/init.cpp]: remove the "Primarily for testing"

**Comment by nkostoulas** [comment]
@AstryX After this is finished you'll also need to change ocean-demo and federation demo in a similar way to the changes required in the unit-tests.

Also you'll need to update https://github.com/commerceblock/ocean-api-reference/blob/master/README.md and https://github.com/commerceblock/ocean-api-reference/blob/master/how-to-ocean-config.md with your new config changes, as well as the equivalent documents in https://github.com/commerceblock/docs.

**Comment by tomt1664** [comment]
> @AstryX After this is finished you'll also need to change ocean-demo and federation demo in a similar way to the changes required in the unit-tests. Then I'll merge this.
> 
> Maybe the mapping repo will require some change too @tomt1664 ? Anything else should be good.

No, this shouldn't change anything with the mapping (scriptPubKeys are used to identify policy outputs here).

**Pull request merged**